src/consensus.rs

```diff
@@ -448,144 +448,175 @@ impl Consensus {
-                // We drive Raft every `tick_period_ms`.
+            let updates = self.advance_node(previous_tick, tick_period)?;
+
+            let mut elapsed = previous_tick.elapsed();
+
+            while elapsed > tick_period {
                 self.node.tick();
-                // Try to reapply entries if some were not applied due to errors.
-                let store = self.node.store().clone();
-                let stop_consensus = store.apply_entries(&mut self.node)?;
+
+                previous_tick += tick_period;
+                elapsed -= tick_period;
+            }
+
+            if self.node.has_ready() {
+                let stop_consensus = self.on_ready()?;
+
                 if stop_consensus {
                     return Ok(());
                 }
-            } else {
-                timeout -= d;
-            }
-            let stop_consensus = self.on_ready()?;
-            if stop_consensus {
-                return Ok(());
+            } else if updates == Some(0) {
+                self.try_sync_local_state()?;
             }
         }
     }
 
-    fn try_sync_local_state(&mut self) -> anyhow::Result<()> {
-        if !self.node.has_ready() {
-            // No updates to process
-            let store = self.node.store();
-            let pending_operations = store.persistent.read().unapplied_entities_count();
-            if pending_operations == 0 && store.is_leader_established.check_ready() {
-                // If leader is established and there is nothing else to do on this iteration,
-                // then we can check if there are any un-synchronized local state left.
-                store.sync_local_state()?;
+    fn advance_node(
+        &mut self,
+        previous_tick: Instant,
+        tick_period: Duration,
+    ) -> anyhow::Result<Option<usize>> {
+        if previous_tick.elapsed() >= tick_period {
+            return Ok(None);
+        }
+
+        if self
+            .try_promote_learner()
+            .context("failed to promote learner")?
+        {
+            return Ok(Some(1));
+        }
+
+        let mut updates = 0;
+        let mut timeout_at = previous_tick + tick_period;
+
+        loop {
+            let message = match self.recv_update(timeout_at) {
+                Ok(message) => message,
+                Err(_) => break,
+            };
+
+            let is_conf_change = matches!(
+                message,
+                Message::FromClient(
+                    ConsensusOperations::AddPeer { .. } | ConsensusOperations::RemovePeer(_)
+                ),
+            );
+
+            if let Err(err) = self.advance_node_impl(message) {
+                log::warn!("{err}");
+                continue;
+            }
+
+            updates += 1;
+            timeout_at = Instant::now() + tick_period / 10;
+
+            if previous_tick.elapsed() >= tick_period || updates >= 128 || is_conf_change {
```

Comment:
Is 128 a batch size?

Reply:
In some sense? `advance_node` does two things: it *propose* new messages to consensus and it *steps* the node with incoming messages from the leader. So this does not directly correlate with how many messages would be send back from this node.

My logic here was:
- we want to tick the node more-or-less reliably every 100ms
- we also want to push as much data into consensus at all times
- every `propose` and every `step` potentially generates some work to be done during `on_ready`
- if we push too many operations (e.g., the node is behind on consensus, and there are a 1000 messages in the `advane_node` queue, and we push all of them into `RawNode`), we might generate too much work to be done in the next 100ms
- so this 128 makes sure we call `on_ready` at least every 128 "events" (propose or step)